File: houseRobber2.js
class Solution {
  /**
   * @param {number[]} nums
   * @return {number}
   */
  rob(nums) {
    const robLine = (nums) => {
      let prePreMax = 0;
      let preMax = 0;
      let trueMax = 0;

      for (const num of nums) {
        trueMax = Math.max(prePreMax + num, preMax);
        prePreMax = preMax;
        preMax = trueMax;
      }

      return trueMax;
    };

    if (nums.length === 1) {
      return nums[0];
    }

    return Math.max(robLine(nums.slice(0, -1)), robLine(nums.slice(1)));
  }
}

File: house_robber_2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        return max(nums[0], self.robLine(nums[:-1]), self.robLine(nums[1:]))

    def robLine(self, nums):
        prev_prev_max, prev_max, true_max = 0, 0, 0

        for i in range(len(nums)):
            true_max = max(prev_max, nums[i] + prev_prev_max)
            prev_prev_max = prev_max
            prev_max = true_max

        return true_max
